source/manual/testing.html.md

```diff
@@ -0,0 +1,91 @@
+
+- **Probes**. Smokey is run in a loop, [with individual alerts for each feature](https://github.com/alphagov/govuk-puppet/blob/c939308b87d9db1128218c951a8c106038c01f50/hieradata_aws/production.yaml#L475). Some of the probes are too flakey to consider paging people, but ideally this is what we should do for each of the features.
+
+- **Surrogate Tests**. [A subset of the probes](https://github.com/alphagov/smokey/blob/main/docs/tagging.md#app-app_name) are run as part of the [Continuous Deployment pipeline](https://docs.publishing.service.gov.uk/manual/development-pipeline.html). These aren’t normal tests as they run in a shared environment and may fail due to unrelated problems.
+
+[We use Sentry to monitor flakey probes](https://sentry.io/organizations/govuk/issues/?project=6370326) and identify patterns we can fix. As GOV.UK shifts towards more instantaneous infrastructure, we should also be able to move towards more pre-merge end-to-end testing and away from reusing Smokey for this purpose.
+
+[Read more about when and how to write Smoke tests](https://github.com/alphagov/smokey/blob/main/docs/writing-tests.md).
+
+## Application Healthchecks
+
+Health checks are a fast way to check an app is running at the end of a deployment. They can also perform detailed checks on the infrastructure the app needs to run.
+
+We have two types of deployment healthcheck:
+
+- Web apps have a [`/healthcheck/ready` endpoint](https://github.com/alphagov/govuk_app_config/blob/main/docs/healthchecks.md), which directly checks if the app can serve requests and connect to its infra e.g. a database.
+
+- Service apps have no web server to query, so [we check the process is running](https://github.com/alphagov/govuk-puppet/blob/32c1bbbb10067078c1406170666a135b4a10aaea/modules/govuk/files/usr/local/bin/govuk_supervised_initctl); apps should fail to start if they can’t connect to their infra ([example](https://github.com/alphagov/email-alert-service/commit/cc6123e88d1ec9542d0d19e719e8e02c37b78538)).
+
+Apps that are automatically deployed all the way to Production should follow the [guidance on healthchecks ("safety checks") in the Continuous Deployment RFC](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md#check-app-is-healthy-ie-it-can-run-in-a-production-environment):
+
+- The healthcheck covers connectivity to all systems the app uses to read or write data e.g. databases, remote file systems, remote caches.
+
+As GOV.UK shifts towards more instantaneous infrastructure, we should be able to run these detailed infrastructure checks as part of pre-merge, end-to-end testing, so that we can learn about configuration issues earlier in the development process.
+
+See also: [how we reuse health check endpoints to alert about transient failures](https://docs.publishing.service.gov.uk/manual/alerts/app-healthcheck-not-ok.html).
+
+## GOV.UK End-to-end Tests
```

Comment:
I struggled with this and Smokey, which are arguably both end-to-end tests.

And it does seem that some people are confused about E2E vs e.g. integration tests.

I'm wondering about the following organisation:

- Site-wide End-to-end Tests (after deployments)
  - Smokey
- Site-wide End-to-end Tests (before merge):
  - Publishing End-to-end Tests

Using "site-wide" instead of GOV.UK should address the ambiguity. We could arguably keep "Smoke Tests", given their multiple uses; but I'm conscious we may be giving a misleading picture of E2E tests overall.

What do you think?

Reply:
Yup, you're right that they are both end-to-end tests. I guess the differentiating factor is the isolated test environment of the publishing end-to-end tests. 

I'm not 100% sold on site-wide as I feel like it implies a wider breadth of testing than we actually do. 

I wonder if you want to just rename "GOV.UK End-to-end tests" as "Publishing end-to-end tests" so it references the product described. I think you're caught on a bit of tightrope where you're trying to explain end-to-end as an abstract concept to allow for future growth in the area, but then by specifying pre-merge you're essentially forced into describing a current part of the system as we don't really know if any future end-to-end tests that could exist will be pre-merge.

Reply:
Yes, exactly. I've added a bit more content to try and make this distinction more explicitly in https://github.com/alphagov/govuk-developer-docs/pull/3564/commits/fc236138676b5982fe1d91e82781f8ba588a8f00 and using the organisation I suggested in https://github.com/alphagov/govuk-developer-docs/pull/3564#discussion_r899980498.